### PR TITLE
Refactored simulation function use the solver_parameters as is instead of unpacking.

### PR DESCRIPTION
**Pull request by dseidita1919**
With the current implementation of the simulation function, the solver_parameters dictionary must be unpacked into the function call in order to run the code with all the arguments being static with respect to the jit. Being able to pass the full dictionary as is would allow the code to be more robust when adding new parameters to the dictionary / codebase as well as make it simpler for users to interface with the code.

The solution written into the pull request is not the only possible solution. What is written is - to my knowledge - the simplest solution which retains the user experience almost exactly. The downside is that it uses global variables to keep the state of the solver_parameters dictionary. The try, except part of the code is there for the first call of the simulation. It generates a simulation function call where the solver_parameters are frozen in. This is repeated in the case that the current solver_parameters are different from the previous (the else branch in the try statement). The if branch is there to prevent jitting the code again if the solver_parameters are unchanged. This prevents the re-jitting overhead at least from call to call.

Other possible solutions to this problem are potentially more pythonic. For example, the simulation function could be separated into two functions:
- a small preprocessing  part of the code
- and the jitted part of the code where the computations take place
This would make it possible for the user to interact with the solver_parameters as a dictionary, but it would just move the dictionary problem to the second portion of the function, so I don't know how strong this solution would be.

Another solution would be to change the simulation function into a simulation class. I think this is probably the most pythonic solution of the 3. This would allow the solver_parameters to be treated as part of the simulation's state and would make the code far more robust to adding new functionality down the line. The downside here is that it would
- require the largest refactor of the code out of these options
- have the largest impact on the user experience (or so I believe)

Rogerio also mentioned that a version of this last solution would be through equinox (in the jax ecosystem). I'm not personally familiar with equinox, but it may be an idea worth looking into if it will be a cleaner way to make this change.

I think that making a change to allow the code to work directly with the dictionary is a positive one for the robustness of the code and making the user experience somewhat more straightforward. I'm hoping for some input here as to what solution would be preferred or if this change seems unnecessary.

### CODE REVIEW

**Comment by dseidita1919** [comment]
@Longyu975 

What are your thoughts on me making an update like this to the code? I personally feel that making the simulation into a class rather than a function is the most Pythonic solution, but I'd like to get your take before moving forward on anything.

**Comment by Longyu975** [comment]
Sorry for the late response, and thanks a lot for the current update!

I also agree that turning the simulation into a class could be a cleaner and more maintainable approach, especially as the number of parameters grows with further development. That said, it really depends on whether you feel it’s worth the effort to refactor. I’m happy either way!

**Review by Longyu975 — APPROVED** [review]

**Review by rogeriojorge — APPROVED** [review]
Looks good to me! We can try to remove the global variable in the future, but for now this seems to fix the issue with minimal changes.

File: jaxincell/_simulation.py [hunk 1]
```diff
@@ -149,9 +151,45 @@ def initialize_simulation_parameters(user_parameters={}):
...
+    """
+    default_solver_parameters = {
+        "number_grid_points": 100,
+        "number_pseudoelectrons": 3000,
+        "number_pseudoparticles_species": None,
+        "total_steps": 1000,
+        "field_solver": 0,
+        "time_evolution_algorithm": 0,
+        "max_number_of_Picard_iterations_implicit_CN": 20,
+        "number_of_particle_substeps_implicit_CN": 2
```
[the remark this answers is not in the document]
Comment (rogeriojorge) [reply jaxincell/_simulation.py]: @copilot apply changes based on [this feedback](https://github.com/uwplasma/JAX-in-Cell/pull/33#discussion_r3066611486)

File: jaxincell/_simulation.py [hunk 2]
```diff
@@ -149,9 +151,45 @@ def initialize_simulation_parameters(user_parameters={}):
 
     return parameters
 
-def initialize_particles_fields(input_parameters={}, number_grid_points=50, number_pseudoelectrons=500,
-                                number_pseudoparticles_species=None, total_steps=350
-                                ,max_number_of_Picard_iterations_implicit_CN=20, number_of_particle_substeps_implicit_CN=2):
+def initialize_solver_parameters(solver_parameters={}):
+    """
```
[the remark this answers is not in the document]
Comment (rogeriojorge) [reply jaxincell/_simulation.py]: @copilot apply changes based on [this feedback](https://github.com/uwplasma/JAX-in-Cell/pull/33#discussion_r3066611467)

**Comment by rogeriojorge** [comment]
@dseidita1919 seems like tests weren't running due to the new syntax. Can you check that after these latest commits, all runs well with pytest?

**Review by rogeriojorge — APPROVED** [review]

**Comment by dseidita1919** [comment]
@Longyu975 @rogeriojorge, new class based architecture is now in and ready for preliminary review. Still need updated unit tests, updated examples, etc. in order to make this complete.

A big thing I'd like is to see what you think about the new structure I pushed for the toml file structure. The main concept is to make it more modular for changes down the road. Looking forward to seeing what you both think about the changes in general and finding all the broken pieces I've missed thus far. More changes to come.

**Comment by dseidita1919** [comment]
@Longyu975 @rogeriojorge Just made some more changes to the Simulation object and _simulation.py file. Should be a bit cleaner now and the implementation should be a little better from the autodiff side. Added some new files to keep everything clean and relatively shorter file length. Everything still needs testing, unit tests, back testing, etc. but should be approaching something that can go to production.

**Review by rogeriojorge — COMMENTED** [review]
Thank you for working on this! Here are a few comments:
- Seems like the _species_parameters.py file is extremely long and verbose. Also the _species_definitions.py file is extremely long. Seems like these changes are going to substantially increase the number of lines in the code, making it probably more difficult to maintain in the future.
- Will the code coverage drop a lot? Are there many untested functions? Right now, JAX-in-Cell has a code coverage of 96%. It would be good to keep it above 95%
- The examples aren't working yet. As an example, I get an error when running the simple Landau_damping.py

 python Landau_damping.py 
Traceback (most recent call last):
  File "/Users/rogeriojorge/local/JAX-in-Cell/examples/Landau_damping.py", line 37, in <module>
    output = block_until_ready(simulation(input_parameters, solver_parameters))
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: Non-hashable static arguments are not supported. An error occurred while trying to hash an object of type <class 'dict'>, {'field_solver': 0, 'number_grid_points': 32, 'number_pseudoelectrons': 40000, 'total_steps': 300, 'time_evolution_algorithm': 0, 'max_number_of_Picard_iterations_implicit_CN': 20, 'number_of_particle_substeps_implicit_CN': 1}. The error was:
TypeError: unhashable type: 'dict'

A more detailed set of comments and suggestions in the following comment.

**Comment by rogeriojorge** [comment]
I agree with the main goal: it would be nice not to require `simulation(input_parameters, **solver_parameters)`, and it would be good to make the distinction between differentiable/runtime inputs and non-differentiable/static configuration clearer.

My concern is that every time we add/change a parameter, we may now need to update several places: defaults, `ALL_*`, `DIFFERENTIABLE_*`, routing, cleaning, hashing, species definitions, legacy mappings, etc. That seems likely to become difficult to maintain.

- Do we really need separate `ALL_DOMAIN_PARAMETERS`, `DIFFERENTIABLE_DOMAIN_PARAMETERS`, and `clean_and_initialize_domain_parameters` objects/functions for every section? It may be simpler to have one source of truth per section. For example:

  DOMAIN_DEFAULTS = {
      "total_steps": 350,
      "timestep_over_spatialstep_times_c": 1.0,
      "number_grid_points": 50,
      "number_grid_points_y": 0,
      "number_grid_points_z": 0,
      "length": 1e-2,
      "length_y": 0,
      "length_z": 0,
      "particle_BC_left": 0,
      "particle_BC_right": 0,
      "field_BC_left": 0,
      "field_BC_right": 0,
  }

  DIFFERENTIABLE_DOMAIN_PARAMETERS = {
      "timestep_over_spatialstep_times_c",
      "length",
      "length_y",
      "length_z",
  }

  ALL_DOMAIN_PARAMETERS = set(DOMAIN_DEFAULTS)


Then `ALL_DOMAIN_PARAMETERS` is derived from the defaults instead of being manually maintained. This would already remove one duplicated list per section. A later improvement could be a small `ParameterSpec` object, but even this small change would make the code less verbose and harder to accidentally break.

- So I think the PR is currently between two APIs: some code now wants `Simulation(parameters).run(...)`, but the examples still call `simulation(input_parameters, solver_parameters)`. We should probably choose one interface and update all examples/tests consistently.

I think the simplest user-facing API would be something like:

  parameters = load_parameters("input.toml")
  sim = Simulation(parameters)
  output = sim.run()

  and for differentiability:

  sim = Simulation(parameters)
  def objective(electron_drift_speed_x):
      return sim.run({
          "electron_drift_speed_x": electron_drift_speed_x,
      })["electric_field"].mean()

  In this model, `Simulation(parameters)` contains the static/non-differentiable configuration, and `run(...)` only accepts differentiable runtime overrides. That seems easier to explain than routing many dictionaries internally.

- I would suggest not passing a raw dictionary as a static argument to a jitted function. That is what causes the non-hashable static argument error above. Either the static configuration should be converted to a hashable structure before JIT, or the jitted function should receive only the actual array/scalar leaves that JAX needs to trace. Longer term, Equinox’s `filter_jit` / pytree filtering might be a cleaner way to separate array/differentiable leaves from static leaves, but I do not think we need to add that dependency in this PR unless we want to.

- It would be good to add tests for at least:

  * `load_parameters(...)`
  * `Simulation(parameters).run()`
  * backwards compatibility with existing input files
  * each example running as a smoke test
  * runtime differentiable parameters working with `jax.grad`
  * static parameters being rejected or handled clearly when passed to `run(...)`
  * species parameters, including legacy species parameters

- My suggestion for a smaller version of this PR would be:

  1. Make the examples work first with one consistent API.
  2. Keep `Simulation(parameters).run(input_parameters=None)` as the main interface.
  3. Avoid passing raw dictionaries as static JIT arguments.
  4. Reduce the parameter metadata duplication by deriving `ALL_*` from the defaults.
  5. Add tests before expanding the species/section routing machinery further.

**Comment by dseidita1919** [comment]
@Longyu975 @rogeriojorge Just made changes to increase the readability of the code - largely as per Rogerio's suggestions. The examples and tests should all now be working as well. Still needs testing to keep the code coverage up, but the codebase as a whole should be starting to look a little better hopefully.

Potential future readability modification / make it easier to add on later is to separate the electrons and ions definitions from _species_definitions.py into their own files. This would be more files but should help with readability and expandability to neutrals in the future - if we want to support them at some point or if we care enough to make them their own thing as opposed to ions with 0 charge. It would mean more places to go if you want to change defaults, though, as I've merged some of the ions and electron defaults as of this moment.

In any case, take a look and let me know if you have any suggestions or comments.

**Comment by dseidita1919** [comment]
New tests are scaffolded - covering the new code and old code with light coverage. Some tests are implemented, but most still need implementation. Aside from that, I think this PR is largely ready for final review from @Longyu975 and @rogeriojorge when you get the chance. I'll keep working on tests, but I'm not planning to touch the rest unless one of you asks for changes.

**Comment by Longyu975** [comment]
Thanks for the update! I tested the examples, and they're working now. The new structure looks really nice overall. So yes, once the test implementation is finished, I think it should be good to go.

File: tests/test_particles.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
+def test_fields_to_particles_grid_interpolates_with_boundary_conditions():
```
Comment (rogeriojorge) [inline tests/test_particles.py:29]: Most of these testes appear to just be scaffolds. Can you remove all scaffold tests to leave the repository clean?

File: tests/test_diagnostics.py [hunk 4]
```diff
@@ -163,3 +163,33 @@ def integrate(y, dx_val):
         assert s["velocities"].shape[0] == T
         assert s["positions"].shape[2] == 3
         assert s["velocities"].shape[2] == 3
+
+
+def test_diagnostics_external_field_energy_nonzero_cases():
```
Comment (rogeriojorge) [inline tests/test_diagnostics.py:215]: Also scaffold

File: tests/test_algorithms.py [hunk 5]
```diff
@@ -150,13 +241,44 @@ def test_boris_step_relativistic_and_field_solver_branch():
     assert rho_step.shape == (G,)
 
 
+def test_boris_step_nonrelativistic_branch_and_zero_field_solver_path():
```
Comment (rogeriojorge) [inline tests/test_algorithms.py:506]: Also scaffold

File: jaxincell/_utils.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
+import jax.numpy as jnp
```
Comment (rogeriojorge) [inline jaxincell/_utils.py]: Just to make sure, are these functions used in multiple files? If they are used only in one file, then we could save one extra  _utils.py file and have these functions appear only in that file.

File: jaxincell/_simulation.py [hunk 7]
```diff
@@ -1,650 +1,442 @@
...
 try: import tomllib
 except ModuleNotFoundError: import pip._vendor.tomli as tomllib
 
 config.update("jax_enable_x64", True)
 
-__all__ = ["initialize_simulation_parameters", "initialize_particles_fields", "simulation", "load_parameters"]
+__all__ = ["Simulation", "load_parameters"]
 
 def load_parameters(input_file):
-    """
```
Comment (rogeriojorge) [inline jaxincell/_simulation.py:23]: Perhaps we can keep docstrings like this one

File: jaxincell/_simulation.py [hunk 8]
```diff
@@ -1,650 +1,442 @@
...
-        # "external_magnetic_field": jnp.array([external_B_field_x, jnp.zeros((number_grid_points,)), jnp.zeros((number_grid_points,))]).T,
-        "external_magnetic_field": parameters["external_magnetic_field"],
-        "number_pseudoelectrons": number_pseudoelectrons,
-        "number_grid_points": number_grid_points,
-        "plasma_frequency": plasma_frequency,
-        "max_initial_vth_electrons": vth_electrons,
-        "max_number_of_Picard_iterations_implicit_CN": max_number_of_Picard_iterations_implicit_CN,
-        "number_of_particle_substeps_implicit_CN": number_of_particle_substeps_implicit_CN,
-    })
+    def classify_and_sort_input_parameters(self, parameters):
```
Comment (rogeriojorge) [inline jaxincell/_simulation.py]: This is a long function in the main file of the program, making a newcomer have to sort through a lot of code to get to the main simulation function. Any way of making this path shorter? This classify_and_sort_input_parameters seems to be the largest function that makes this path difficult. On another note, perhaps the _simulation function can be moved to the top of the file, if it's possible to make it easier to be found?

File: jaxincell/_simulation.py [hunk 9]
```diff
@@ -1,650 +1,442 @@
...
-    else:
-        initial_carry = (
-            E_field, B_field, positions,
-            velocities, qs, ms, q_ms,
+        self.fields = field_state["fields"]
+        self.external_magnetic_field = field_state["external_magnetic_field"]
+        self.external_electric_field = field_state["external_electric_field"]
+
+    @partial(jit, static_argnames=['self', 'domain_hash', 'species_hash', 'external_field_hash', 'source_hash', 'solver_hash'])
+    def _simulation(self, input_parameters=None, domain_hash='', species_hash='', external_field_hash='', source_hash='', solver_hash=''):
```
Comment (rogeriojorge) [inline jaxincell/_simulation.py:130]: These arguments 

domain_hash='', species_hash='', external_field_hash='', source_hash='', solver_hash=''

are never called or used in the code. Why are they here? Can they be removed?

**Review by rogeriojorge — CHANGES_REQUESTED** [review]
See comments inline. Two additional points:
- I tried running the examples but they don't print to the terminal the same information that was before. Example on the previous version with the Weibel instability example:

```
python Weibel_instability.py 
Length of the simulation box: 165.0 Debye lengths or 11.667261889578036 Skin Depths
Density of electrons: 4.2712375916956744e+16 m^-3
Electron temperature: 2554.994758743706 eV
Ion temperature / Electron temperature: 1
Debye length: 0.001818181818181818 m
Skin depth: 0.025712973861328998 m
Wavenumber * Debye length: 0.0
Pseudoparticles per cell: 20.0
Pseudoparticle weight: 4271237591695.674
Steps at each plasma frequency: 12.8564869306645
Total time: 194.45436482630058 / plasma frequency
Number of particles on a Debye cube: 256723521.6646535
Relativistic gamma factor: Maximum 1.0427176228099204, Average 1.0012364749817266
Charge x External electric field x Debye Length / Temperature: 0.0

Running for 2,500 iterations: 100%|██████████████████████████████████████████████████████████| 2500/2500 [00:13<00:00, 186.35it/s]
```

while the new code prints only

```
python Weibel_instability.py                                    
Running for 2,500 iterations: 100%|██████████████████████████████████████████████████████████| 2500/2500 [00:09<00:00, 262.88it/s]
```

- The examples between the main branch and ds/solver_parameters_to_dictionary do not output the same. Electric and magnetic fields, as well as phase space, are different.

Image

Image
Image
Image

File: tests/test_particles.py (hunk 3 above)
Comment (dseidita1919) [reply tests/test_particles.py:29]: Planning to implement scaffolds before merging.

File: jaxincell/_utils.py (hunk 6 above)
Comment (dseidita1919) [reply jaxincell/_utils.py]: Yes, these functions are used in at least two files. Previously they were defined in one and then imported by another. I can swap back or define in both if you'd like.

File: jaxincell/_simulation.py (hunk 9 above)
Comment (dseidita1919) [reply jaxincell/_simulation.py:130]: These are here to make sure that Jax recompiles whenever changes are made to the parameters. Each hash is recreated whenever a change is made (see the getter and setter functions below). It isn't a unique solution, but it is a solution that allows you to swap back and forth between two configurations without recompiling when you revisit a set of parameters.

File: jaxincell/_utils.py (hunk 6 above)
Comment (dseidita1919) [reply jaxincell/_utils.py]: Actually, it looks like a lot of those references became unused at some point. Only as_float_parameter() gets called in multiple files and it's only really a return statement. I'll go ahead and make some changes to get rid of this file.

File: jaxincell/_simulation.py (hunk 8 above)
Comment (dseidita1919) [reply jaxincell/_simulation.py]: I've made a change to make this section a little shorter and cleaner. I'll also look at moving _simulation up to the top.

**Comment by dseidita1919** [comment]
I can't say with certainty until I go in and check more thoroughly, but I suspect the differences in the phase space and fields are due to initial positions and velocities rather than a difference in the function of the code. That would reflect the random seeds being handled differently which lines up with the minimal changes I made to the loop and larger changes I made to the initialization. I'll try to confirm this first and if the code is generating different seeds I'll fix the random seeds such that the examples line up with the existing examples.

As for the printing, I'll go ahead and add those print statements back in. I'll check across all the examples to see if they went away in other places also.

**Comment by dseidita1919** [comment]
Found the printing problem! In rearranging functions a print statement in the original initialize_particles_fields() function never found a home in any of the new initialization functions. Fixing that now.

**Comment by dseidita1919** [comment]
The difference in the outputs between main and this branch was actually twofold! First, this branch was handling the rng seeds for every species differently from how main handles them. Secondly, there is actually a mistake in main's input.toml file. The input.toml file on main has
"ion_temperature_over_electron_temperature = 0.01"
but this doesn't actually set any of the x, y, or z temperature ratios to 0.01. Thus, it's actually running with temperature ratios of 1. In any case, my plots now align with the plots you posted above from the main branch. I'll push the changes with the fixed rng and the temperature ratio change shortly.

**Comment by dseidita1919** [comment]
Just pushed changes, so the code should be fully comparable at this point @rogeriojorge . Please let me know if there are any other issues. Aside from that, just need to keep implementing tests.

**Comment by dseidita1919** [comment]
The tests in test_sources.py (the code I just pushed and can remove if the tests end up being poorly done) are testing functions I'm less familiar with, so I would appreciate if @Longyu975 or @rogeriojorge could review the tests to make sure Codex created reasonable tests which are correctly and accurately testing the functions in _sources.py.

**Review by rogeriojorge — COMMENTED** [review]
This is looking much better now! A few final cleanup items:

1. CI needs to pass before merge (check the comment below).

2. I'm fine if the scaffold tests are removed for now and we focus on tests later (JAX-in-Cell wasn't actually fully tested before this merge)

3. There are still mutable default arguments, for example `Simulation.__init__(self, parameters={})` and several cleaner functions such as `clean_and_initialize_domain_parameters(..., input_parameters={})`. Isn't it a better practice to change those to `None` defaults?

4. The hash arguments in `_simulation` are still a little non-obvious:
   ```python
   domain_hash='', species_hash='', external_field_hash='', source_hash='', solver_hash=''
````

I understand that they are used to force JAX recompilation when static config changes, but since they are not used inside the function body, please add a short comment above `_simulation` explaining that they are intentionally static cache keys. That will prevent future maintainers from deleting them as “unused”.

5. Optional but worthwhile: keep a minimal docstring or short comment near the main class or functions, explaining the new mental model:

   * `Simulation(parameters)` is static/non-differentiable configuration
   * `sim.run(input_parameters)` is for differentiable runtime overrides only


------------

One final comment on the CI: it looks like a Python/JAX compatibility issue rather than a code failure. The workflow still tests Python 3.8, but `jaxlib` no longer has a compatible wheel available for Python 3.8, so `pip install .` fails before tests run.

Can we make this small metadata/CI update?

# .github/workflows/build_test.yml
- python-version: [ '3.8', '3.9', '3.10', '3.11', '3.12']
+ python-version: [ '3.9', '3.10', '3.11', '3.12']

and

# pyproject.toml
- dependencies = [ "jax", "jaxlib", "jax_tqdm", "matplotlib"]
- requires-python = ">=3.5"
+ dependencies = [ "jax", "jax_tqdm", "matplotlib"]
+ requires-python = ">=3.9"

**Comment by dseidita1919** [comment]
Updated CI, fixed the mutable default arguments, and added a decent amount of documentation to some functions. Also moved _simulation() up near the top of the Simulation object for readability. Added more testing too. @rogeriojorge hopefully this fixes the problems you identified. We'll know where CI lands soon hopefully.

**Review by rogeriojorge — APPROVED** [review]
Looks good! I'm approving it for merging. Some minor cleanup points if you don't mind while we're doing this merge:
- The pyproject.toml file still only knows about

[tool.setuptools]
packages = ["jaxincell"]

But this PR adds and imports a real subpackage, jaxincell._parameters, with its own __init__.py. The source-tree tests can pass because pytest runs from the repo checkout, but a built wheel may omit jaxincell/_parameters, causing installed users to hit an import error.
I would change the packaging before merge:

[tool.setuptools.packages.find]
include = ["jaxincell*"]

or, minimally:

[tool.setuptools]
packages = ["jaxincell", "jaxincell._parameters"]

The find version is better because future subpackages will be included automatically.

I would also add a tiny CI check after pip install .:

- name: Check installed package import outside source tree
  run: |
    cd /tmp
    python -c "import jaxincell; import jaxincell._parameters; from jaxincell import Simulation"

That catches exactly this class of issue.

2. External fields look disabled in the Boris step

In _algorithms.py, the code currently has:

total_E = E_field #+ external_field_parameters["external_electric_field"]
total_B = B_field #+ external_field_parameters["external_magnetic_field"]

Given that the PR still initializes external electric and magnetic field arrays in initialize_field_state, and returns them from the field state, the Boris pusher should use them.

I would either fix it before merge or explicitly remove the external-field argument from Boris_step until it is supported. My preference is to fix it now, because the old code added external fields into the particle pusher path.

Probably the cleaner pattern is to pass the initialized arrays, not the raw parameter dict:

external_E = field_state["external_electric_field"]
external_B = field_state["external_magnetic_field"]

step_func = lambda carry, step_index: Boris_step(
    carry, step_index, solver_parameters,
    external_E, external_B,
    dx, dt, grid, box_size,
    particle_BC_left, particle_BC_right,
    field_BC_left, field_BC_right,
    solver_parameters["field_solver"],
)

and then:

total_E = E_field + external_E
total_B = B_field + external_B

Add one small test: with zero self-fields and a nonzero external electric field, a particle’s velocity changes in the expected direction. That would make this safe.

- There seems to be a misleading class docstring example:

input_parameters = Simulation.input_parameters
grad(sim.run)(input_parameters)

That should probably be:

input_parameters = sim.input_parameters

and the grad example should wrap sim.run(...) in a scalar objective, because sim.run returns a dictionary. Not a blocker, but it is a quick fix.

**Comment by dseidita1919** [comment]
Your suggestions should now be implemented along with reimplementing the user overrides of the positions and velocities. At this point I think this branch matches the capabilities of main, and all the tests are implemented. Someone other than me should double check the tests at some point, but once CI is good on this most recent push (ended up having to go with ["jaxincell", "jaxincell._parameters"] as ["jaxincell*"] caused CI to fail), I think we're good to finish this PR.
